nuxeo-features/nuxeo-annotation/nuxeo-annotation-core/src/main/java/org/nuxeo/ecm/annotation/AnnotationServiceImpl.java

```diff
@@ -0,0 +1,137 @@
+package org.nuxeo.ecm.annotation;
+
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_COLOR_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_CREATION_DATE_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_DATE_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_DOCUMENT_ID_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_DOCUMENT_XPATH_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_DOC_TYPE;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_FLAGS_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_LAST_MODIFIER_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_NAME_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_OPACITY_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_PAGE_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_POSITION_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_SECURITY_PROPERTY;
+import static org.nuxeo.ecm.annotation.AnnotationConstants.ANNOTATION_SUBJECT_PROPERTY;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.platform.query.api.PageProvider;
+import org.nuxeo.ecm.platform.query.api.PageProviderService;
+import org.nuxeo.ecm.platform.query.nxql.CoreQueryAndFetchPageProvider;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.DefaultComponent;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * @since 10.1
+ */
+public class AnnotationServiceImpl extends DefaultComponent implements AnnotationService {
+
+    private static final Log log = LogFactory.getLog(AnnotationServiceImpl.class);
+
+    protected static final String ANNOTATION_NAME = "annotation";
+
+    protected static final String ANNOTATION_PAGEPROVIDER_NAME = "GET_ANNOTATIONS_FOR_DOCUMENT";
+
+    @Override
+    public Annotation createAnnotation(CoreSession session, String documentId, String documentXpath,
+            Annotation annotation) {
+
+        // Create annotation as a placeless document
+        DocumentModel annotationModel = session.createDocumentModel(null, ANNOTATION_NAME, ANNOTATION_DOC_TYPE);
+        annotationModel.setPropertyValue(ANNOTATION_DOCUMENT_ID_PROPERTY, documentId);
+        annotationModel.setPropertyValue(ANNOTATION_DOCUMENT_XPATH_PROPERTY, documentXpath);
+        setAnnotationProperties(annotationModel, annotation);
+        annotationModel = session.createDocument(annotationModel);
+        return new AnnotationImpl(annotationModel);
```

Comment:
Would it make sense to have `Annotation` contributed as an `adapter` to the `org.nuxeo.ecm.core.api.DocumentAdapterService` extension point so we could use `annotationModel.getAdapter(Annotation.class)` instead of an instantiation coupled to the `AnnotationImpl` implemenation?

This would be consistent with the common use of adapters in Nuxeo to map a `DocumentModel` to a POJO and allow customizing the Annotation factory.

Reply:
The way we map a `DocumentModel` to an `Annotation` is an internal detail of the annotation service. There's no reason to let external callers know about it or do it. Callers will always pass a POJO AnnotationImpl. I'm not even sure we should make the `AnnotationImpl(DocumentModle)`constructor public.